src/parser/shared/normalizers/PrePullCooldowns.js

```diff
@@ -0,0 +1,172 @@
+          if (buffSpells[i].buffId === event.ability.guid) {
+            debug && console.debug(`Detected a precast buff cooldown: ${event.ability.name}`);
+            prepullCasts.push(this.constructor._fabricateCastEvent(event, buffSpells[i].castId));
+            buffSpells.splice(i, 1);
+            break;
+          }
+        }
+        continue;
+      }
+
+      if (event.type === 'cast') {
+
+        /*
+         * This will copy the first resource information to all precast events.
+         * It's not pretty or 100% accurate, but it prevents errors on analyzers
+         * that require resource information and usually wont be too far off.
+         *
+         * In the future, a better way to handle this would be to make sure any
+         * analyzers that need the resource information check for .prepull, or
+         * we gain access to spell data that lets us build the resource
+         * information more accurately.
+         */
+        if (precastClassResources === null && event.classResources) {
+          debug && console.debug("Setting prepull class resources to:", event.classResources);
+          precastClassResources = event.classResources;
+        }
+
+        // If a cast is found for a damage spell, remove it from the search
+        for (let i = 0; i < damageSpells.length; i += 1) {
+          if (damageSpells[i].castId === event.ability.guid) {
+            damageSpells.splice(i, 1);
+            break;
+          }
+        }
+        continue;
+      }
+
+      if (event.type === 'damage') {
+        // If a damage event already has a cast event, it shouldn't be in the array
+        for (let i = 0; i < damageSpells.length; i += 1) {
+          if (damageSpells[i].damageIds.some(id => id === event.ability.guid)) {
+            debug && console.debug(`Detected a precast damage cooldown: ${event.ability.name}`);
+            prepullCasts.push(this.constructor._fabricateCastEvent(event, damageSpells[i].castId));
+            damageSpells.splice(i, 1);
+            break;
+          }
+        }
+      }
+
+    }
+
+    /*
+     * Potential issue: If players cast buffs long before the pull that have a
+     * long duration, this normalizer doesn't know that and assumes they
+     * cast it optimized right before the pull. This means that the next time
+     * it gets cast we may get an error saying it was still on cooldown. To
+     * resolve this, we could also listen for buffremove events and use that
+     * to deduce the cast time. This would require ability info about durations.
+     */
+    // Working backwards, assume worst-case GCD timestamps
+    let totalGCD = 0;
+    for (let i = prepullCasts.length - 1; i >= 0; i -= 1) {
+      const event = prepullCasts[i];
+      const ability = this.abilities.getAbility(event.ability.guid);
+      const gcd = (ability && ability.gcd && ability.gcd.base) || 1500;
+      totalGCD += gcd;
+      event.timestamp = firstTimestamp - totalGCD;
```

Comment:
Because this is a normalizer I don't have access to a lot of those properties that rely on analyzers (haste, gcd, etc.). That was one of the hurdles that led to a delay in this PR, deciding what to do about those issues. For now I decided it was best to just assume worst-case GCD and use base *if* it was there. But now that you've pointed it out I've forgotten that base can be a func as well. So I guess I should just check for both and see if one isn't a func, and use that?